File: 2020/day15/solution1.py
from argparse import ArgumentParser

def read_file(input_example: bool):
    fname = 'input_example.txt' if input_example else 'input1.txt'
    with open(fname) as f:
        return [int(el) for el in f.read().split(",")]



class GameRunner:

    def __init__(self, data, turn_end=2020):
        self.turn_end = turn_end
        self.number_last_two_spoken = {}
        for (i, el) in enumerate(data):
            self.number_last_two_spoken[el] = [i+1, -1]
        self.last_number = data[-1]
        self.turn = len(data) + 1

    def has_been_spoken(self, number):
        return number in self.number_last_two_spoken and self.number_last_two_spoken[number][1] != -1

    def last_two_call_diff(self, number):
        t1, t2 = self.number_last_two_spoken[number]
        return t1 - t2

    def run(self):
        while self.turn <= self.turn_end:
            if not self.has_been_spoken(self.last_number):
                next_number = 0
            else:
                next_number = self.last_two_call_diff(self.last_number)
            if next_number not in self.number_last_two_spoken:
                self.number_last_two_spoken[next_number] = [self.turn, -1]
            else:
                last_turn_number = self.number_last_two_spoken[next_number][0]
                self.number_last_two_spoken[next_number] = [self.turn, last_turn_number]
            self.last_number = next_number
            self.turn += 1
        
        return self.last_number


if __name__ == "__main__":
    parser = ArgumentParser()
    parser.add_argument('--input-example', '-i', type=eval, choices=[True, False], default=False, dest='input_example')
    input_example = parser.parse_args().input_example
    data = read_file(input_example)
    g = GameRunner(data)
    result = g.run()
    print(result)

File: 2020/day15/solution1.rb
require 'set'
require_relative 'parser'

class GameRunner

    def initialize(data, turn_end=2020)
        @turn_end = turn_end
        @number_last_two_spoken = {}
        data.each_with_index do |el, idx|
            @number_last_two_spoken[el] = [idx+1, -1]
        end
        @last_number = data[-1]
        @turn = data.length + 1
    end

    def has_been_spoken(number)
        @number_last_two_spoken.keys.include?(number) and @number_last_two_spoken[number][1] != -1
    end

    def last_two_call_diff(number)
        t1, t2 = @number_last_two_spoken[number]
        t1 - t2
    end

    def run
        while @turn <= @turn_end
            if not has_been_spoken @last_number
                next_number = 0
            else
                next_number = last_two_call_diff @last_number
            end
            if not @number_last_two_spoken.keys.include? next_number
                @number_last_two_spoken[next_number] = [@turn, -1]
            else
                last_turn_number = @number_last_two_spoken[next_number][0]
                @number_last_two_spoken[next_number] = [@turn, last_turn_number]
            end
            @last_number = next_number
            @turn += 1
        end
        @last_number
    end

end


data = read_file
g = GameRunner.new(data)
puts g.run
